### ISSUE TITLE
clarification on the understanding of losses

### ISSUE THREAD

**Issue by stormxuwz** [inclusionAI/TwinFlow#5]
Hi, very interesting paper and nice work!! Would you mind my asking a few questions regarding the paper?

1. Regarding self-adversarial loss and rectification loss, 
 
It seems the basic idea is to train a single lora for the one base model that can capture both targeting distillation real velocity and fake velocity (on the results predicted by the model with negative t, instead of training another lora on another base model). And rectification loss is actually the DMD loss that uses the fake velocity trained with self-adversarial objective. is my understanding correct? 

2. In section 3.3, "practical implementation of mixed loss", I had a hard time understanding why  "L_twinflow necessitates a fixed target time of r = 0". Would you mind giving a little bit more details?

Looking forward to your kind reply and the training code open-sourced!

**Comment by QAQdev** [comment]
Thanks for the interest in our work!

> It seems the basic idea is to train a single lora for the one base model that can capture both targeting distillation real velocity and fake velocity (on the results predicted by the model with negative t, instead of training another lora on another base model).

Yes. **One of our core contribution is integrating the function of one-step generator, real velocity, and fake velocity into one model.** For the experiments using LoRA, there is only one LoRA model that acts as one-step generator, real velocity, and fake velocity. It is worth noting that we also conducted full-parameter training on Qwen-Image (c.f. "Further exploration on 20B full-parameter training on Qwen-Image."), and the performance is better in this case (the open-source model we provide is also obtained through full-parameter training).

> And rectification loss is actually the DMD loss that uses the fake velocity trained with self-adversarial objective. is my understanding correct?

They are not exactly the same. The DMD loss generally comprises a distribution matching loss (see Algorithm 1 in VSD [1]) and a regression loss. In our method, we do not have a regression loss.

Furthermore, in terms of the distribution matching loss, both VSD and DMD aim to minimize the discrepancy between the real and fake scores to zero. _**Our objective is to minimize the discrepancy between the real and fake flow velocity fields to zero.**_ Such a difference leads to different final distribution matching losses (see Eq. 6 in our paper and Eq. 7 in the DMD paper).

_**Additionally, DMD and VSD use a designed weighting function for the distribution matching loss, while we do not apply any weighting strategy in the distribution matching loss to keep the formulation clean.**_

> In section 3.3, "practical implementation of mixed loss", I had a hard time understanding why "L_twinflow necessitates a fixed target time of r = 0".

Because of $L_{base}$ we use (which comes from RCGM [2]). As defined in Eq. 1 in our paper, the base loss in TwinFlow is an *any-step loss*, conditioned on both the current time $t$ and the target time $r$. $L_{base}$ learns any $t$ to $r$ transitions on PF-ODE. To help understanding, for example, MeanFlow also learns any-step, Eq. 4-9 in their paper also derive how the model learns any-step transitions.

_(Note: We selected the RCGM loss as our base loss for two reasons: (1) calculating the RCGM loss does not require Jacobian-Vector Product (JVP) computation, whereas MeanFlow's loss function does, which significantly increases GPU memory usage; and (2) 2nd-order or 3rd-order RCGM exhibits high stability during training, whereas MeanFlow is typically unstable for large models.)_

The reason we use fixed target time of $r = 0$ is to construct the prediction ($x^{fake}$) of the network which points to the endpoint on the PF-ODE, i.e. $r=0$. Here, The general form is $x^{fake} =  x_{t}^{real} -  F_{\theta}(x_{t}^{real}, r) \cdot t$, where target time $r=0$. In our method and experiments, $t=1$, thus $x^{fake} = z - F_{\theta}(z, 0)$, which is a one-step prediction of the network. $x^{fake}$ will then be used for the rectification part to let the model learn one-step generation.

We are organizing our code, stay tuned!

[1] VSD, https://arxiv.org/abs/2305.16213
[2] RCGM, https://github.com/LINs-lab/RCGM/blob/main/assets/paper.pdf

**Comment by QAQdev** [comment]
@stormxuwz 
We have provided a tutorial on MNIST to help better understand the core implementation, you can find it here: https://github.com/inclusionAI/TwinFlow?tab=readme-ov-file#tutorials

**Comment by stormxuwz** [comment]
Thank you so much for the detail explanations! I will looking through all the referred paper and the tutorial. Nice work!

**Comment by yyfcc17** [comment]
i have a question about equation 6.

if i get it right, equation 6 aims to transform the score in the distribution matching loss to velocity of flow models.

you derived a `-(1-t) / t` to rectify the grad for flow matching, why this recitification term is not used in equation 9 and your code?

if this recitification term is taken out, the grad is computed the same way as DMD for score matching.

thank you.

**Comment by sp12138** [comment]
Hi, @yyfcc17 

Thank you for your interest in our work!

We observed that DMD/VSD-style score matching can be interpreted as our "velocity matching" scaled by a specific time-dependent weight.

We chose to exclude this weight for two reasons:

It adds unnecessary complexity to both the formulation and the code.

Empirically, direct velocity matching (without the weight) outperforms the weighted score matching approach in our experiments.

Therefore, strictly following the unweighted velocity matching formulation proved to be both simpler and more effective.

Please feel free to reach out if you have any further questions!
